up_client_socket/rust/u_transport_socket.rs

```diff
@@ -0,0 +1,395 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2023 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use async_std::io;
+use async_trait::async_trait;
+use tokio::runtime::Runtime;
+use up_rust::UListener;
+use up_rust::{UAttributesValidators, UriValidator};
+use up_rust::{UCode, UMessage, UMessageType, UStatus, UTransport, UUri};
+
+use protobuf::Message;
+use std::net::TcpStream as TcpStreamSync;
+use std::{
+    collections::HashMap,
+    sync::{Arc, Mutex},
+};
+use std::io::{Read, Write};
+
+use crate::constants::BYTES_MSG_LENGTH;
+use crate::constants::DISPATCHER_ADDR;
+
+pub trait UtransportExt {
+     fn socket_init(&mut self);
+     fn _handle_publish_message(&mut self, umsg: UMessage);
+     fn _handle_request_message(&mut self, umsg: UMessage);
+    fn read_socket(&self, buffer: &mut [u8]) -> io::Result<usize>;
+}
+
+pub struct UtransportSocket {
+    socket_sync: TcpStreamSync,
+    listner_map: Arc<Mutex<HashMap<String, Vec<Arc<dyn UListener>>>>>,
+}
+impl Clone for UtransportSocket {
+    fn clone(&self) -> Self {
+        UtransportSocket {
+            socket_sync: self
+                .socket_sync
+                .try_clone()
+                .expect("issue in cloning sync socket"),
+            listner_map: self.listner_map.clone(),
+        }
+    }
+}
+
+impl UtransportSocket {
+    pub fn new() -> Self {
+        let socket_sync: TcpStreamSync =
+            TcpStreamSync::connect(DISPATCHER_ADDR).expect("issue in connecting  sync socket");
+
+        UtransportSocket {
+            socket_sync,
+            listner_map: Arc::new(Mutex::new(HashMap::new())),
+        }
+    }
+}
+
+impl UtransportExt for UtransportSocket {
+     fn socket_init(&mut self) {
+        loop {
+            // Receive data from the socket
+            let mut buffer: [u8; BYTES_MSG_LENGTH] = [0; BYTES_MSG_LENGTH];
+
+            let bytes_read = match self.read_socket(&mut buffer) {
+                Ok(bytes_read) => bytes_read,
+                Err(e) => {
+                    dbg!("Socket error: {}", e);
+                    break;
+                }
+            };
+
+            // Check if no data is received
+            if bytes_read == 0 {
+                continue;
+            }
+
+            let umessage = UMessage::parse_from_bytes(&buffer[..bytes_read]).expect("Failed to parse message");
+
+            match umessage.attributes.type_.enum_value() {
```

Comment:
This would make it a little cleaner imho
```suggestion
            match umessage.attributes.type_.enum_value_or(UMESSAGE_TYPE_UNSPECIFIED) {
                UMESSAGE_TYPE_UNSPECIFIED => {}
                UMessageType::UMESSAGE_TYPE_PUBLISH => {}
                UMessageType::UMESSAGE_TYPE_REQUEST => {}
                UMessageType::UMESSAGE_TYPE_RESPONSE => {}
                UMessageType::UMESSAGE_TYPE_NOTIFICATION => {}
            }
```

Reply:
corrected.